source/sonata_tech.rst

```diff
@@ -576,3 +576,23 @@ Required fields for ``Spykfunc``:
 .. _specification: https://github.com/AllenInstitute/sonata/blob/master/docs/SONATA_DEVELOPER_GUIDE.md
 .. _enumeration: https://github.com/AllenInstitute/sonata/blob/master/docs/SONATA_DEVELOPER_GUIDE.md#nodes---enum-datatypes
 .. _Tsodyks Markram Model: https://www.pnas.org/content/94/2/719
+
+Format of the electrodes_file
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The format of the weights file is described below:
+
+.. table::
+
+  ============================================= ======================= ========== =================== ============= ==========================================
+  Group                                         Field                   Type       Shape               Requirement   Description
+  ============================================= ======================= ========== =================== ============= ==========================================
+  /electrodes/{electrodename}                   layer                   utf8       1                   Optional      Layer of the circuit in which {electrodename} is located. If the electrode is in a region without cortical layers, then "NA". If the electrode is outside the brain, then "Outside"
+  /electrodes/{electrodename}                   position                float32    3                   Mandatory     Position of {electrodename} in microns, in cartesian coordinates
+  /electrodes/{electrodename}                   region                  utf8       1                   Optional      Region in which {electrodename} is located
+  /electrodes/{electrodename}                   type                    utf8       1                   Optional      Either EEG or LFP
+  /electrodes/{electrodename}/{population_name} index                   uint64     1                   Mandatory     Index of the column corresponding to this electrode in /electrodes/{population_name}/scaling_factors
```

Comment:
If the cardinality of the electrodes is low, is there a need for the indirection from this `index` to the column in `/electrodes/{population_name}/scaling_factors`; why not have a 1-d dataset under the `/electrodes/{electrodename}/{population_name}/scaling_factors`}

Reply:
could you give an example?

Reply:
In `source/usecases/usecase1/electrodes_file/electrode_file.h5` currently:
```
HDF5 "source/usecases/usecase1/electrodes_file/electrode_file.h5" {
DATASET "/electrodes/Neuropixels-12_0/nodeA/index" {
   DATATYPE  H5T_STD_I64LE
   DATASPACE  SCALAR
   DATA {
   (0): 0
   }
}
```
Signifying using the column 0 of `/electrodes/nodeA/scaling_factors Dataset {2887, 12}`, correct?

Why not drop `/electrodes/nodeA/scaling_factors`, and directly put the 0th column under
`/electrodes/Neuropixels-12_0/nodeA/scaling_factors`?

Reply:
This would be inefficient to read since we want all the electrodes for each compartment (basically the row in the scaling_factors)

Reply:
How many electrodes are generally being created?

Reply:
Aren't the `Total_comp` pretty high (like O(1e9))? I'm surprised making one read vs. 384 would make that much difference, considering how much data will be coming across the line.

Reply:
Yes but in neurodamus you go gid by gid, it doesnt make sense to query all the electrode groups one by one when you can access the offsets for the gid and then get the subset from the scaling_factors directly

Reply:
I'm not sure I follow; are you loading the data from the h5 all at once, or piece by piece on each query time?

I did a quick bench with:
```
import h5py
import numpy as np

def build(path, shape):
    data = np.random.random(shape)
    with h5py.File(path, 'w') as h5:
        h5.create_dataset('matrix', data=data)
        for i in range(shape[1]):
            h5.create_dataset(str(i), data=data[:, i])


def read_as_columns(path, shape):
    data = {}

    with h5py.File(path, 'r') as h5:
        for i in range(shape[1]):
            data[i] = h5[str(i)][:]

    return data

def read_as_matrix(path):
    with h5py.File(path, 'r') as h5:
        return h5['matrix'][:]

path = 'test.h5'
electrode_count = 400
shape = (1_000_000, electrode_count)

build(path, shape)
%timeit read_as_columns(path, shape)
%timeit read_as_matrix(path)
```
And I'm not seeing a large difference in loading time.

Reply:
i load all at once for each gid:

```
def get_node_id_subsets(self, node_id, population_name):
        node_ids = self._lfp_file[population_name]["node_ids"]
        # Look for the index of the node_id
        index = numpy.where(numpy.array(node_ids) == node_id)[0][0]
        offsets_dataset = self._lfp_file[population_name]["offsets"]
        electrodes_dataset = self._lfp_file["electrodes"][population_name]["scaling_factors"]
        index_low = offsets_dataset[index]
        index_high = offsets_dataset[index + 1]
        # Get the subsets for the data for the node_id
        subset_data = electrodes_dataset[index_low:index_high, :]
```

Reply:
I see; I assume the plan is to move this all into libsonata once the standard is set, right?

Reply:
My idea was to have only the field for the electrodes_file (see here: https://github.com/BlueBrain/libsonata/pull/271) and that code to stay in neurodamus since the info about the population is there

Reply:
> since the info about the population is there

Not sure what you mean; I thought that the viz team was also going to have to read this, sometimes, to make illustrations?

Reply:
But this would be for the metadata right? as I understand the data itself it will only be used by neurodamus to generate the LFP report

Reply:
Not sure, generally things that are added to the spec get an implementation in libsonata.  In this case, I could see that it might be useful to be able to visualize the scaling factors on the actual morphologies.

As long as it doesn't impact the data layout, I don't think it matters, and we can cross that bridge in the future.

Reply:
I have created a ticket to implement the reading of the electrodes file in libsonata
https://bbpteam.epfl.ch/project/issues/browse/BBPBGLIB-1053

I think we should merge this and work on the libsonata implementation whenever we have resources available.
@jamesgking @pramodk @jdcourcol